_posts/2025-02-28-a-practical-guide-for-selecting-HNSW-hyperparameters.md

```diff
@@ -0,0 +1,683 @@
+| `w_recall`     | 0.9 | 0.8 | 0.7 | 0.6 | 0.5 |
+| `w_throughput` | 0.1 | 0.2 | 0.3 | 0.4 | 0.5 |
+
+## Evaluation  
+
+We evaluated our method in two scenarios:  
+
+1) **Leave-one-out evaluation** – One of the 15 datasets is used as the test dataset while the remaining datasets serve as the training set.  
+2) **Deployment evaluation** – All 15 datasets are used for training, and the method is tested on four additional datasets using a new embedding model, [Cohere-embed-english-v3](https://huggingface.co/Cohere/Cohere-embed-english-v3.0), which was not part of the training set.  
+
+The first scenario mimics cross-validation in machine learning, while the second simulates an evaluation in which the complete training dataset is used for model deployment.  
+
+### Leave-one-out evaluation  
+
+For this evaluation, we determined the ground-truth configurations under different weightings by applying our method to the test dataset. We then compared these with the predicted configurations derived from the training datasets using the same method.  
+
+We calculated the mean absolute error (MAE) between the predicted and ground-truth configurations for normalized (min-max scaled) recall and throughput. The following bar plot shows the average MAE across all 15 datasets in the leave-one-out evaluation:  
+
+![MAE compared with ground truth](/assets/media/blog-images/2025-02-28-a-pratical-guide-for-selecting-HNSW-hyperparameters/mae.png){:class="img-centered"}  
+
+The results show that the average MAEs for normalized recall are below 0.1. For context, if recall values range from 0.5 to 0.95 on a dataset, an MAE of 0.1 translates to a raw recall difference of only 0.045. This suggests that the predicted configurations closely match the ground-truth configurations, particularly for high-recall weightings.  
+
+The MAEs for throughput are larger, likely because throughput measurements tend to be noisier than recall measurements. However, the MAEs decrease when higher weight is assigned to throughput.  
+
+### Deployment evaluation  
+
+For this evaluation, we applied our method to the 15 training datasets and tested the resulting configurations on three datasets using the Cohere-embed-english-v3 embedding model. Our goal was to ensure that the learned configurations align with the Pareto front, representing different trade-offs between recall and throughput.  
+
+The following plot depicts the recall and throughput for the learned configurations in different colors, with other configurations displayed in gray.  
+
+![Trade-off of the 5 configurations](/assets/media/blog-images/2025-02-28-a-pratical-guide-for-selecting-HNSW-hyperparameters/tradeoff.png){:class="img-centered"}  
+
+The results show that the five selected configurations effectively cover the high-recall and high-throughput regions. Since we did not assign high weights to throughput, the learned configurations do not extend into the low-recall, high-throughput area.
```

Comment:
```suggestion
The results show that the five selected configurations effectively cover the high-recall and high-throughput regions. Because we did not assign high weights to throughput, the learned configurations do not extend into the low-recall, high-throughput area.
```